Add how to create a fungible token (#397)
onflow/docs: docs/guides/fungible-token.md

---
title: How to Create a Fungible Token on Flow
sidebar_label: Create a Fungible Token
description: Guide to creating a fungible token on Flow with the Flow CLI and Cadence.
sidebar_position: 4
------

This article will guide you through the steps to create, deploy, mint, and transfer fungible tokens on Flow.

## What are Fungible Tokens?

Fungible tokens are digital assets that are interchangeable and indistinguishable with other tokens of the same type. This means that each token is identical in specification to every other token in circulation. Think of them like traditional money; every dollar bill has the same value as every other dollar bill. Fungible tokens play a crucial role in web3 ecosystems, serving as both a means of payment and an incentive for network participation. They can take on various roles including currencies, structured financial instruments, shares of index funds, and even voting rights in decentralized autonomous organizations.

## Vaults on Flow

> On the Flow blockchain, fungible tokens are stored in structures called vaults. Think of a vault as a digital piggy bank. When you transfer tokens from one vault to another:

Reviewer: I would recommend linking to the cadence tutorial or something that goes into a little more detail about how resources are used to represent vaults on Flow for tokens

Reviewer 2: Feel free to add a PR for this update @joshuahannan. Would be much appreciated!

            let vault <- from as! @FooToken.Vault
            emit TokensDeposited(amount: vault.balance, to: self.owner?.address)
            self.balance = self.balance + vault.balance
            vault.balance = 0.0
            destroy vault
        }

        pub fun withdraw(amount: UFix64): @Vault {
            self.balance = self.balance - amount
            emit TokensWithdrawn(amount: amount, from: self.owner?.address)
            return <- create Vault(balance: amount)
        }

        destroy() {
            FooToken.totalSupply = FooToken.totalSupply - self.balance
        }

        init(balance: UFix64) {
            self.balance = balance
        }
    }

    pub resource Minter {
        pub fun mintToken(amount: UFix64): @FungibleToken.Vault {
            FooToken.totalSupply = FooToken.totalSupply + amount
            return <- create Vault(balance: amount)
        }

        init(){}
    }

    pub fun createEmptyVault(): @FooToken.Vault {
        return <- create Vault(balance: 0.0)
    }

    init() {
        self.totalSupply = 1000.0
        self.account.save(<- create Minter(), to: /storage/Minter)
    }
}
```

## Deploying the Contract

> In order to use the contract, we need to deploy it to the network we want to use it on. In our case we are going to deploy it to emulator while developing.

Reviewer 2: ```suggestion
In order to use the contract, we need to deploy it to the network we want to use it on. In this case we are going to deploy it to emulator while developing. Alternatively we could configure `testnet` or `mainnet`.
```

```
"deployments": {
  "emulator": {
    "emulator-account": ["FooToken"]
  }
}
```

Next, using the Flow CLI, we will start the emulator. As mentioned, this will give us a local development environment for the Flow Blockchain.

```
flow emulator start
```

Open a new terminal and run the following to deploy your project:

```
flow project deploy
```

Congrats, you've deployed your contract to the Flow Blockchain emulator. To read more about deploying your project to other environments, see the [CLI docs](https://developers.flow.com/tools/flow-cli/deployment/deploy-project-contracts).

## Reading the Token’s Total Supply

Let's now check that our total supply was initialized with 1,000 FooTokens. Go ahead and create a new directory and script file called `TotalSupply.cdc`.

```
mkdir cadence/scripts
touch cadence/scripts/TotalSupply.cdc
```

In `TotalSupply.cdc`, let's add this code which will log the `totalSupply` value from the `FooToken` contract:

```
import "FooToken"

pub fun main() {
  log(FooToken.totalSupply)
}
```

To run this using the CLI, enter this in your terminal:

```
flow scripts execute cadence/scripts/TotalSupply.cdc
```

> In the terminal where you started the emulator, you should see 1,000.

Reviewer 2: ```suggestion
In the terminal where you started the emulator, you should see `1000.0`
```

Let's first create the file for this transaction called `CreateVault.cdc`:

```
mkdir cadence/transactions
touch cadence/transactions/CreateVault.cdc
```

Then add this code to it. This will call the `createEmptyVault` function, save it in storage, and create a capability for the vault which will later allow us to read from it (To learn more about capabilities, see [the Cadence docs here](https://developers.flow.com/cadence/language/capabilities)).

```
import "FooToken"
import "FungibleToken"

transaction {
	prepare(acct: AuthAccount) {
	  acct.save(<- FooToken.createEmptyVault(), to: /storage/Vault)
	  acct.link<&FooToken.Vault{FungibleToken.Balance, FungibleToken.Receiver}>(/public/Vault, target: /storage/Vault)
	}

	execute {
		log("Created")
	}
}
```

> Let's next create a new emulator account using the CLI. We'll use this account to create a new vault and mint tokens into it. Run:

Reviewer 2: ```suggestion
Next let's create a new emulator account using the CLI. We'll use this account to create a new vault and mint tokens into it. Run:
```

```

Next, let's add the following code to the `Minter.cdc` file. This code will attempt to borrow the minting capability and mint 20 new tokens into the receivers account.

```
import "FooToken"
import "FungibleToken"

transaction(receiverAccount: Address) {
	prepare(acct: AuthAccount) {
	  let minter = acct.borrow<&FooToken.Minter>(from: /storage/Minter)
		  ?? panic("Can't borrow Minter")

		let newVault <- minter.mintToken(amount: 20.0)

		let receiverVault = getAccount(receiverAccount).getCapability(/public/Vault)
			.borrow<&FooToken.Vault{FungibleToken.Receiver}>()
			?? panic("Could not get public Vault")

		receiverVault.deposit(from: <- newVault)
	}

	execute {}
}
```

To run this transaction, enter this in your terminal. Note: `123` should be replaced with address of `test-acct` found in your `flow.json`. This command also states to sign with our `emulator-account` on the Emulator network.

```
flow transactions send ./cadence/transactions/Minter.cdc 123 --signer emulator-account --network emulator
```

Let's go ahead and read the vault again. Remember to replace `123` with the correct address.

```
flow scripts execute cadence/scripts/ReadVaultBalance.cdc 123
```

> It should now say 20 are in the vault.

Reviewer 2: ```suggestion
It should now say 20 tokens are in the vault.
```

```

Let's add the code which states that the signer of the transaction will withdraw from their vault and put it into the receiver's vault which will be passed as a transaction argument.

```
import "FooToken"
import "FungibleToken"

transaction(receiverAccount: Address, amount: UFix64) {
	prepare(acct: AuthAccount) {
	  let signerVault = acct.borrow<&FooToken.Vault>(from: /storage/Vault)
		  ?? panic("Couldn't get signer's Vault")

	let receiverVault = getAccount(receiverAccount).getCapability(/public/Vault)
			.borrow<&FooToken.Vault{FungibleToken.Receiver}>()
			?? panic("Could not get public Vault")

	  receiverVault.deposit(from: <- signerVault.withdraw(amount: amount))
	}

	execute {
		log("Transferred")
	}
}
```

To send our tokens, we'll need to create a new account to send them to. Let's make one more account on emulator. Run:

```
flow accounts create
```

And pick the name:

```
test-acct-2
```

Make sure to select Emulator as the network.

Don't forget the new account will need a vault added, so let's run the following transaction to add one:

```
flow transactions send ./cadence/transactions/CreateVault.cdc --signer test-acct-2 --network emulator
```

> Now, let's send one token from our earlier account to the new account. Remember to replace `123` with account address of `test-acct-2`.

Reviewer 2: ```suggestion
Now, let's send 1 token from our earlier account to the new account. Remember to replace `123` with account address of `test-acct-2`.
```

```
flow transactions send ./cadence/transactions/Transfer.cdc 123 "1.0" --signer test-acct --network emulator
```

After that, read the balance of `test-acct-2` (replace the address `123`).

```
flow scripts execute cadence/scripts/ReadVaultBalance.cdc 123
```

> You should now see one token in `test-acct-2` account!

Reviewer 2: ```suggestion
You should now see 1 token in `test-acct-2` account!
```
